File: src/main/java/me/zeroeightsix/kami/command/commands/XRayCommand.java
package me.zeroeightsix.kami.command.commands;

import me.zeroeightsix.kami.command.Command;
import me.zeroeightsix.kami.command.syntax.ChunkBuilder;
import me.zeroeightsix.kami.command.syntax.parsers.EnumParser;
import me.zeroeightsix.kami.module.modules.render.XRay;
import net.minecraft.block.Block;

import static me.zeroeightsix.kami.util.text.MessageSendHelper.sendChatMessage;
import static me.zeroeightsix.kami.util.text.MessageSendHelper.sendWarningMessage;

public class XRayCommand extends Command {
    public XRayCommand() {
        super("xray", new ChunkBuilder().append("command", true, new EnumParser(new String[]{"help", "list", "+block", "-block", "=block", "defaults", "clear", "invert"})).build());
        setDescription("Allows you to add or remove blocks from the &fxray &7module");
    }

    @Override
    public void call(String[] args) {
        if (!XRay.INSTANCE.isEnabled()) {
            sendWarningMessage("&6Warning: The " + XRay.INSTANCE.getName().getValue() + " module is not enabled!");
            sendWarningMessage("These commands will still have effect, but will not visibly do anything.");
        }
        for (String s : args) {
            if (s == null)
                continue;
            if (s.equalsIgnoreCase("help")) {
                sendChatMessage("The " + XRay.INSTANCE.getName().getValue() + " module has a list of blocks");
                sendChatMessage("Normally, the " + XRay.INSTANCE.getName().getValue() + " module hides these blocks");
                sendChatMessage("When the Invert setting is on, the " + XRay.INSTANCE.getName().getValue() + " only shows these blocks");
                sendChatMessage("This command is a convenient way to quickly edit the list");
                sendChatMessage("Available options: \n" +
                        "+block: Adds a block to the list\n" +
                        "-block: Removes a block from the list\n" +
                        "=block: Changes the list to only that block\n" +
                        "list: Prints the list of selected blocks\n" +
                        "defaults: Resets the list to the default list\n" +
                        "clear: Removes all blocks from the " + XRay.INSTANCE.getName().getValue() + " block list\n" +
                        "invert: Quickly toggles the invert setting");
            } else if (s.equalsIgnoreCase("clear")) {
                XRay.INSTANCE.extClear();
                sendWarningMessage("Cleared the " + XRay.INSTANCE.getName().getValue() + " block list");
            } else if (s.equalsIgnoreCase("defaults")) {
                XRay.INSTANCE.extDefaults();
                sendChatMessage("Reset the " + XRay.INSTANCE.getName().getValue() + " block list to default");
            } else if (s.equalsIgnoreCase("list")) {
                sendChatMessage("\n" + XRay.INSTANCE.extGet());
            } else if (s.equalsIgnoreCase("invert")) {
                if (XRay.INSTANCE.invert.getValue()) {
                    XRay.INSTANCE.invert.setValue(false);
                    sendChatMessage("Disabled " + XRay.INSTANCE.getName().getValue() + " Invert");
                } else {
                    XRay.INSTANCE.invert.setValue(true);
                    sendChatMessage("Enabled " + XRay.INSTANCE.getName().getValue() + " Invert");
                }
            } else if (s.startsWith("=")) {
                String sT = s.replace("=", "");
                XRay.INSTANCE.extSet(sT);
                sendChatMessage("Set the " + XRay.INSTANCE.getName().getValue() + " block list to " + sT);
            } else if (s.startsWith("+") || s.startsWith("-")) {
                String name = s.substring(1);
                Block b = Block.getBlockFromName(name);
                if (b == null) {
                    sendChatMessage("&cInvalid block name <" + name + ">");
                } else {
                    if (s.startsWith("+")) {
                        sendChatMessage("Added <" + name + "> to the " + XRay.INSTANCE.getName().getValue() + " block list");
                        XRay.INSTANCE.extAdd(name);
                    } else {
                        sendChatMessage("Removed <" + name + "> from the " + XRay.INSTANCE.getName().getValue() + " block list");
                        XRay.INSTANCE.extRemove(name);
                    }
                }
            } else {
                sendChatMessage("&cInvalid subcommand <" + s + ">");
            }
        }
    }
}

File: src/main/java/me/zeroeightsix/kami/command/commands/XRayCommand.kt
package me.zeroeightsix.kami.command.commands

import me.zeroeightsix.kami.command.ClientCommand
import me.zeroeightsix.kami.module.modules.render.XRay
import me.zeroeightsix.kami.util.text.MessageSendHelper
import me.zeroeightsix.kami.util.text.formatValue

// TODO: Remove once GUI has List
object XRayCommand : ClientCommand(
    name = "xray",
    description = "Manage xray blocks"
) {
    init {
        literal("add", "+") {
            block("block") { blockArg ->
                execute("Add a block to xray list") {
                    val blockName = blockArg.value.registryName.toString()

                    XRay.INSTANCE.extAdd(blockName)
                    MessageSendHelper.sendChatMessage("${formatValue(blockName)} has been added to the xray block list")
                }
            }
        }

        literal("remove", "-") {
            block("block") { blockArg ->
                execute("Remove a block from xray list") {
                    val blockName = blockArg.value.registryName.toString()

                    XRay.INSTANCE.extRemove(blockName)
                    MessageSendHelper.sendChatMessage("Removed ${formatValue(blockName)} from xray block list")
                }
            }
        }

        literal("set", "=") {
            block("block") { blockArg ->
                execute("Set the xray list to one block") {
                    val blockName = blockArg.value.registryName.toString()

                    XRay.INSTANCE.extSet(blockName)
                    MessageSendHelper.sendChatMessage("Set the xray block list to ${formatValue(blockName)}")
                }
            }
        }

        literal("list") {
            execute("Print xray block list") {
                MessageSendHelper.sendChatMessage("The following blocks are set in xray:")
                MessageSendHelper.sendRawChatMessage(XRay.INSTANCE.extGet())
            }
        }

        literal("reset", "default") {
            execute("Reset the xray list to defaults") {
                XRay.INSTANCE.extDefaults()
                MessageSendHelper.sendChatMessage("Reset the xray block list to defaults")
            }
        }

        literal("clear") {
            execute("Set the xray list to nothing") {
                XRay.INSTANCE.extClear()
                MessageSendHelper.sendChatMessage("Cleared the xray block list")
            }
        }

        literal("invert") {
            execute("Invert xray block list") {
                XRay.INSTANCE.invert.value = !XRay.INSTANCE.invert.value
                MessageSendHelper.sendChatMessage("Inverted the xray block list")
            }
        }
    }
}